Refactoring, add-marks, add-search

Keindel opened the pull request:
- Refactoring:

1. add-marks
2. refactor rating and popularity
3. refactor recommendations
4. refactor popular by director
5. add-tests

- add-search

Keindel committed:
fix: sorting in getDirectorFilms

Keindel committed:
Add search (#23)

Add search

Keindel committed:
Merge remote-tracking branch 'origin/develop' into develop

Keindel committed:
fix

Keindel committed:
fix

Commit:
fix: entityId (#24)

fix: entityId

Keindel committed:
Add marks (#20)

feat: change likes to marks
feat: adapt-recommendations-algorithm
fix: change the algorithm of the most popular films
feat: add rating
fix: changed the sorting of films from likes to ratings
feat: add tests

Keindel committed:
fix + style

Keindel committed:
style

Rexsus-bit committed:
Test for refactoring (#28)

add Test for refactoring

AliKamil commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -117,4 +123,7 @@ public Collection<Film> getSortedFilms(long directorId, String sortBy) throws Va
         return filmStorage.getDirectorFilms(directorId, sortBy);
     }
 
+    public Collection<Film> getSearch(String query, String by) {
```
[Minor] Bad naming, use `search` or `queryBy` instead.

AliKamil commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -51,11 +51,11 @@ public void requestFriendship(Long userId, Long friendToAddId) throws UserNotFou
 
-    public Collection<Long> getMutualFriendsIds(Long user1Id, Long user2Id) throws UserNotFoundException, FilmNotFoundException, MpaNotFoundException, GenreNotFoundException, DirectorNotFoundException {
+    public Collection<Long> getMutualFriendsIds(Long user1Id, Long user2Id) throws UserNotFoundException {
         long userIdWithLessFriends = Long.min(getFriendsNum(user1Id), getFriendsNum(user2Id));
```
This logic has O(n) unnecessary queries to the db. Consider the following:

```
User firstUserFriends = userStorage.getById(firstUserId).getFriends().keySet();
User secondUserFriends = userStorage.getById(secondUserId).getFriends().keySet();

List<Long> commonFriends = firstUserFriends.size() >= secondUserFriends.size() ? firstUserFriends : secondUserFriends;
commonFriends.retainAll(firstUserFriends);
commonFriends.retainAll(secondUserFriends); //awkward, can be optimized.

return commonFriends;
```

Alternatively and potentially quicker would be to make an inner join on DB side.

Keindel replied:
Thank you! Missed this legacy from version with no DB.

Keindel commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -50,20 +50,22 @@ public Film update(@Valid @RequestBody @NonNull Film film) throws FilmValidation
-    public void addLikeFromUser(@PathVariable("id") Long filmId, @PathVariable Long userId) throws UserNotFoundException, FilmNotFoundException {
-        filmService.likeFromUser(filmId, userId);
+    @PutMapping("/{id}/{mark}/{userId}")
+    public void addMarkFromUser(@PathVariable("id") Long filmId, @PathVariable Integer mark, @PathVariable Long userId)
```
Task description explicitly declared to use PUT.
![image](https://user-images.githubusercontent.com/97405919/177985176-bc7ed8b9-bea6-4784-ab45-1fd562ee4620.png)
see Technical Specification per sprint 9

Keindel commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -54,6 +55,10 @@ private void validateFilm(Film film) throws FilmValidationException {
     }
 
+    private void validateMark(int mark) throws MarkValidationException {
+        if (mark < 1 || mark > 10) throw new MarkValidationException();
```
yes

Keindel commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/FilmDbStorage.java:
```diff
@@ -199,59 +207,60 @@ public void deleteById(Long id) throws FilmNotFoundException {
-                " FROM likes" +
-                " WHERE film_id IN " + sqlFilmsOfUser +
+    public Collection<Film> getRecommendedFilmsForUser(Long userId) {
+        String sqlAllMarkedFilmsOfUser = "(SELECT m1.film_id FROM marks AS m1 WHERE mark_from_user = ?)";
```
Thanks!)

Keindel commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/FilmDbStorage.java:
```diff
@@ -302,43 +326,94 @@ public Collection<Film> getDirectorFilms(long directorId, String sortBy) throws
+        Collection<Film> films;
+        List<Long> filmIds;
+        switch (by) {
+            case "director":
```
@Mark33SC

Keindel commented on src/test/java/ru/yandex/practicum/filmorate/TestsForRefactoring.java:
```diff
@@ -0,0 +1,303 @@
+@SpringBootTest
+@AutoConfigureTestDatabase
+@RequiredArgsConstructor(onConstructor_ = @Autowired)
+class TestsForRefactoring {
```
@Rexsus-bit

Keindel commented on src/main/java/ru/yandex/practicum/filmorate/service/FeedService.java:
```diff
@@ -37,9 +31,8 @@ public Collection<Feed> feeds(Long id){
-        UserDbStorage userDbStorage = (UserDbStorage) userStorage;
-        userDbStorage.getWithoutFriendsByIdOrThrowEx(userId);
-        userDbStorage.getWithoutFriendsByIdOrThrowEx(friendToAddId);
+        userStorage.getWithoutFriendsByIdOrThrowEx(userId);
```
@aBoyarov

Keindel commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/FeedDbStorage.java:
```diff
@@ -124,25 +120,22 @@ public void likeFromUser(long filmId, long userId) {
-                friend_id = 2;
-            }
+            user_id = rowSet.getInt("user_id");
             String sqlQuery =
```
@aBoyarov

Keindel committed:
fix per review comments (#30)

Mark33SC committed:
fix duplicate per review (#31)

Commit:
Fix branch (#32)

fix: remove duplicates from FeedDbStorage
fix: remove unnecessary validate user_id in the FeedService

Keindel committed:
feeds.userId not null

Mark33SC committed:
Test for refactoring (#33)

* fix tests per review

Keindel committed:
feeds.userId not null

Keindel committed:
fix exception in addOrRemoveFriend

Keindel committed:
delete reference for feed.userId

AliKamil commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -47,39 +41,26 @@ public User update(User user) throws UserNotFoundException, FilmNotFoundExceptio
     }
 
-    public Collection<Long> getMutualFriendsIds(Long user1Id, Long user2Id) throws UserNotFoundException, FilmNotFoundException, MpaNotFoundException, GenreNotFoundException, DirectorNotFoundException {
-        long userIdWithLessFriends = Long.min(getFriendsNum(user1Id), getFriendsNum(user2Id));
```
Nice, looks so much better now.

AliKamil commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/FeedDbStorage.java:
```diff
@@ -25,43 +26,62 @@ public class FeedDbStorage implements FeedStorage {
-            , "FRIEND"
-            , "ADD"
-            , friendToAddId);
+    public void addFriend(long userId, long friendToAddId) throws UserNotFoundException {
```
Nice, good refactoring.

AliKamil commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/FeedDbStorage.java:
```diff
@@ -79,128 +99,49 @@ public boolean updateFriend(long userId, long friendId) {
-                , "FRIEND"
-                , "UPDATE"
-                , entityId);
+                    , userId
```
Why break line before comma? :)

AliKamil commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/FeedDbStorage.java:
```diff
@@ -212,31 +153,81 @@ public void deleteReview(Long reviewId) {
+
+    private void addOrRemoveMarkFromUser(long userId, long filmId, String action) {
+        Timestamp timestamp = new Timestamp(System.currentTimeMillis());
+        String sqlQuery2 = "select user_id from users where user_id  = ?";
```
If this query is needed to make sure that user with this id exists, we can use a subselect rather than a separate query. Better option would be to let foreign key constraint to take care of this edgecase.

AliKamil reviewed (approved):
Database files should not be commited, add them to gitignore.
